modules/externalipam/netbox/netbox_client.rb

```diff
@@ -0,0 +1,201 @@
+  # Implementation class for External IPAM provider Netbox
+  class NetboxClient
+    include Proxy::Log
+    include Proxy::Ipam::IpamHelper
+    include Proxy::Ipam::IpamValidator
+
+    def initialize(conf)
+      @api_base = "#{conf[:url]}/api/"
+      @token = conf[:token]
+      @api_resource = Proxy::Ipam::ApiResource.new(api_base: @api_base, token: "Token #{@token}")
+      @ip_cache = Proxy::Ipam::IpCache.instance
+      @ip_cache.provider_name('netbox')
+    end
+
+    def get_ipam_subnet(cidr, group_name = nil)
+      if group_name.nil? || group_name.empty?
+        get_ipam_subnet_by_cidr(cidr)
+      else
+        group_id = get_group_id(group_name)
+        get_ipam_subnet_by_group(cidr, group_id)
+      end
+    end
+
+    def get_ipam_subnet_by_group(cidr, group_id)
+      params = URI.encode_www_form({ status: 'active', prefix: cidr, vrf_id: group_id })
+      response = @api_resource.get("ipam/prefixes/?#{params}")
+      json_body = JSON.parse(response.body)
+      return nil if json_body['count'].zero?
+      subnet = subnet_from_result(json_body['results'][0])
+      return subnet if json_body['results']
+    end
+
+    def get_ipam_subnet_by_cidr(cidr)
+      params = URI.encode_www_form({ status: 'active', prefix: cidr })
+      response = @api_resource.get("ipam/prefixes/?#{params}")
+      json_body = JSON.parse(response.body)
+      return nil if json_body['count'].zero?
+      subnet = subnet_from_result(json_body['results'][0])
+      return subnet if json_body['results']
+    end
+
+    def ipam_groups
+      response = @api_resource.get('ipam/vrfs/')
+      json_body = JSON.parse(response.body)
+      groups = []
+
+      return groups if json_body['count'].zero?
+
+      json_body['results'].each do |group|
+        groups.push(
+          {
+            name: group['name'],
+            description: group['description'],
+          }
+        )
+      end
+
+      groups
+    end
+
+    def get_ipam_group(group_name)
+      raise ERRORS[:groups_not_supported] unless groups_supported?
+      response = @api_resource.get("ipam/vrfs/?name=#{group_name}")
+      json_body = JSON.parse(response.body)
+      return nil if json_body['count'].zero?
+
+      group = {
+        id: json_body['results'][0]['id'],
+        name: json_body['results'][0]['name'],
+        description: json_body['results'][0]['description'],
+      }
+
+      return group if json_body['results']
+    end
+
+    def get_group_id(group_name)
+      return nil if group_name.nil? || group_name.empty?
+      group = get_ipam_group(group_name)
+      raise ERRORS[:no_group] if group.nil?
+      group[:id]
+    end
+
+    def get_ipam_subnets(group_name)
+      if group_name.nil?
+        params = URI.encode_www_form({ status: 'active' })
+      else
+        group_id = get_group_id(group_name)
+        params = URI.encode_www_form({ status: 'active', vrf_id: group_id })
+      end
+
+      response = @api_resource.get("ipam/prefixes/?#{params}")
```

Comment:
I would first make a hash of params and then later use that
```suggestion
      params = {status: 'active}
      params[:vrf_id] = get_group_id(group_name) if group_name

       response = @api_resource.get("ipam/prefixes/?#{URI.encode_www_form(params)}") 
```